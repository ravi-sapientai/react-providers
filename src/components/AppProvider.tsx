import * as React from 'react';
import { IContextMap, ContextStatuses, IContext } from '../interfaces';

const _contexts: IContextMap = {};

export const AppProvider = (data: { contexts: IContextMap; children?: JSX.Element }) => {
  const providers: React.ComponentType[] = [];
  const { contexts, children } = data;
  // init context
  Object.keys(contexts).forEach((name) => {
    const { Provider, Consumer } = contexts[name];
    _contexts[name] = { Provider, Consumer, status: ContextStatuses.PENDING };
  });

  // create providers graph
  const resolveDependenciesAndAdd = (context: IContext) => {
    if (context.status === ContextStatuses.INCLUDED) {
      return;
    }
    if (context.status === ContextStatuses.PROCESSING) {
      throw new Error(`Cyclic dependency was detected.`);
    }
    context.status = ContextStatuses.PROCESSING;

    if (context.Provider.contextDependency) {
      context.Provider.contextDependency.forEach((contextName: string) => {
        resolveDependenciesAndAdd(_contexts[contextName]);
      });
    }

    providers.push(context.Provider);
    context.status = ContextStatuses.INCLUDED;
  };

  Object.keys(_contexts).forEach((contextName: string) => {
    resolveDependenciesAndAdd(_contexts[contextName]);
  });

  // check if any contexts were setted
  if (providers.length === 0) {
    return children || null;
  }

  return providers
    .reverse()
    .reduce((store: JSX.Element, Provider: React.ComponentType) => <Provider>{store}</Provider>, children);
};

AppProvider.displayName = 'AppProvider';

// Get contexts to calculate dependencies in use HOC
export const getCurrentContexts = () => {
  return _contexts;
};
