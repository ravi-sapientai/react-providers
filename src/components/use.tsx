import * as React from 'react';
import * as hoistStatics from 'hoist-non-react-statics/dist/hoist-non-react-statics.cjs';
import { getCurrentContexts } from './AppProvider';
import { IComponentAdditionals, IContextMap } from '../interfaces';

export const use = (context: string | Array<string> | IContextMap) => (
  Component: (React.ComponentType) & IComponentAdditionals
) => {
  let contexts: Array<string> | IContextMap = context as Array<string> | IContextMap;
  if (typeof context === 'string') {
    contexts = [context];
  }

  class Wrap extends React.Component {
    static displayName = `use(${Component.displayName || Component.name})`;
    static contextDependency;

    /**
     * Render a react subTree without intermediate components.
     * So we can get a clear subtree (with recursion):
     * ```
     *  <Consumer1>
     *    {value1 =>
     *      <Consumer2>
     *        {value2 =>
     *          <YourWrappedComponent name1={value1} name2={value2} />
     *        }
     *      </Consumer2>
     *    }
     *  </Consumer1>
     * ```
     * instead of (without recursion):
     * ```
     *  <Consumer1>
     *    {value1 =>
     *      <Wrap name1={value1}>
     *        <Consumer2>
     *          {value2 =>
     *            <YourWrappedComponent name1={value1} name2={value2} />
     *          }
     *        </Consumer2>
     *      </Wrap>
     *    }
     *  </Consumer1>
     * ```
     */
    renderPreparedMap(contextNames: Array<string>, contextMap: IContextMap) {
      const propsMap: { [key: string]: any } = {};

      const renderItem = (index: number): JSX.Element => {
        if (index === contextNames.length) {
          return <Component {...this.props} {...propsMap} />;
        }

        const contextName = contextNames[index];
        const Consumer: React.ComponentType = contextMap[contextName].Consumer;
        return (
          <Consumer>
            {(props) => {
              propsMap[contextName] = props;
              return renderItem(index + 1);
            }}
          </Consumer>
        );
      };

      return renderItem(0);
    }

    renderUsingArray(contextArray: Array<string>) {
      const nameWithoutSuchContext = contextArray.find(
        (contextName) => !getCurrentContexts().hasOwnProperty(contextName)
      );
      if (nameWithoutSuchContext) {
        throw new Error(`There is no such context name ${nameWithoutSuchContext}.`);
      }

      return this.renderPreparedMap(contextArray, getCurrentContexts());
    }

    renderUsingDict(contextDict: IContextMap) {
      const propsMap: { [key: string]: any } = {};

      return this.renderPreparedMap(Object.keys(contextDict), contextDict);
    }

    render() {
      if (Array.isArray(contexts)) {
        return this.renderUsingArray(contexts);
      }

      return this.renderUsingDict(contexts);
    }
  }

  // Describe dependencies to automatically resolver
  if (Component.contextDependency && Array.isArray(Component.contextDependency)) {
    if (Array.isArray(contexts)) {
      Wrap.contextDependency = [...Component.contextDependency, ...contexts];
    } else {
      // We detect dependencies between different context using there names
      Wrap.contextDependency = [...Component.contextDependency, ...Object.keys(contexts)];
    }
  } else {
    if (Array.isArray(contexts)) {
      Wrap.contextDependency = contexts;
    } else {
      // We detect dependencies between different context using there names
      Wrap.contextDependency = Object.keys(contexts);
    }
  }

  return hoistStatics(Wrap, Component);
};
