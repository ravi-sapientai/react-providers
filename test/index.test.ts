import './AppProvider.test';
import './use.test';
import './useWithoutAppProvider.test';
