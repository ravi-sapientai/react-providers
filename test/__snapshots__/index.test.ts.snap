// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppProvider> adds Context.Provider when rendering 1`] = `
<AppProvider
  contexts={
    Object {
      "some": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <TodosContextProvider>
    <div />
  </TodosContextProvider>
</AppProvider>
`;

exports[`<AppProvider> renders correctly 1`] = `""`;

exports[`<AppProvider> resolves dependencies in contexts 1`] = `
<AppProvider
  contexts={
    Object {
      "contextWithDependency": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "some": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <TodosContextProvider>
    <use(ContextWithDependencyProvider)>
      <ContextWithDependencyProvider
        some={
          Object {
            "add": [Function],
            "data": Object {
              "list": Array [],
            },
          }
        }
      >
        <div />
      </ContextWithDependencyProvider>
    </use(ContextWithDependencyProvider)>
  </TodosContextProvider>
</AppProvider>
`;

exports[`<AppProvider> works with complex dependencies e.g. A depends on C, B depends both on A and on C 1`] = `
<AppProvider
  contexts={
    Object {
      "a": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "b": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "c": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <CProvider>
    <use(AProvider)>
      <AProvider
        c={
          Object {
            "data": "Hello world",
          }
        }
      >
        <use(BProvider)>
          <BProvider
            a={
              Object {
                "data": "Hello world",
              }
            }
            c={
              Object {
                "data": "Hello world",
              }
            }
          >
            <div />
          </BProvider>
        </use(BProvider)>
      </AProvider>
    </use(AProvider)>
  </CProvider>
</AppProvider>
`;

exports[`<Use> renders correctly 1`] = `
<AppProvider
  contexts={
    Object {
      "some": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <TodosContextProvider>
    <use(WrappedComponent)>
      <WrappedComponent
        some={
          Object {
            "add": [Function],
            "data": Object {
              "list": Array [],
            },
          }
        }
      >
        <div />
      </WrappedComponent>
    </use(WrappedComponent)>
  </TodosContextProvider>
</AppProvider>
`;

exports[`<Use> works with complex dependencies e.g. A depends on C, B depends both on A and on C 1`] = `
<AppProvider
  contexts={
    Object {
      "a": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "b": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "c": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <CProvider>
    <use(AProvider)>
      <AProvider
        c={
          Object {
            "data": "Hello world",
          }
        }
      >
        <use(BProvider)>
          <BProvider
            a={
              Object {
                "data": "Hello world",
              }
            }
            c={
              Object {
                "data": "Hello world",
              }
            }
          >
            <use(use(use(WrappedComponent)))>
              <use(use(WrappedComponent))
                a={
                  Object {
                    "data": "Hello world",
                  }
                }
              >
                <use(WrappedComponent)
                  a={
                    Object {
                      "data": "Hello world",
                    }
                  }
                  b={
                    Object {
                      "data": "Hello world",
                    }
                  }
                >
                  <WrappedComponent
                    a={
                      Object {
                        "data": "Hello world",
                      }
                    }
                    b={
                      Object {
                        "data": "Hello world",
                      }
                    }
                    c={
                      Object {
                        "data": "Hello world",
                      }
                    }
                  >
                    <div />
                  </WrappedComponent>
                </use(WrappedComponent)>
              </use(use(WrappedComponent))>
            </use(use(use(WrappedComponent)))>
          </BProvider>
        </use(BProvider)>
      </AProvider>
    </use(AProvider)>
  </CProvider>
</AppProvider>
`;

exports[`<Use> works with complex dependencies using arrays 1`] = `
<AppProvider
  contexts={
    Object {
      "a": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "b": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
      "c": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "unstable_read": [Function],
        },
        "Provider": [Function],
      },
    }
  }
>
  <CProvider>
    <use(AProvider)>
      <AProvider
        c={
          Object {
            "data": "Hello world",
          }
        }
      >
        <use(BProvider)>
          <BProvider
            a={
              Object {
                "data": "Hello world",
              }
            }
            c={
              Object {
                "data": "Hello world",
              }
            }
          >
            <use(WrappedComponent)>
              <WrappedComponent
                a={
                  Object {
                    "data": "Hello world",
                  }
                }
                b={
                  Object {
                    "data": "Hello world",
                  }
                }
                c={
                  Object {
                    "data": "Hello world",
                  }
                }
              >
                <div />
              </WrappedComponent>
            </use(WrappedComponent)>
          </BProvider>
        </use(BProvider)>
      </AProvider>
    </use(AProvider)>
  </CProvider>
</AppProvider>
`;
